src/types.ts

```diff
@@ -197,3 +197,71 @@ export interface HistoryLookup {
 export const HistoryLookupSpec: ObjectSpec<HistoryLookup> = {
   bugs: array(object(BugHistorySpec)),
 };
+
+export interface Comment {
+  attachment_id: int | undefined;
+  bug_id: int;
+  count: int;
+  creation_time: datetime;
+  creator: string;
+  id: int;
+  is_markdown: boolean;
+  is_private: boolean;
+  tags: string[];
+  time: datetime;
+  text: string;
+}
+
+export const CommentSpec: ObjectSpec<Comment> = {
+  attachment_id: optional(int),
+  bug_id: int,
+  count: int,
+  creation_time: datetime,
+  creator: string,
+  id: int,
+  is_markdown: boolean,
+  is_private: boolean,
+  tags: array(string),
+  time: datetime,
+  text: string,
+};
+
+export interface CommentsTemplate {
+  [key: number]: {
+    comments: Comment | Comment[];
+  };
+}
+
+export const CommentsTemplateSpec: ObjectSpec<CommentsTemplate> = {
+  key: object(CommentSpec),
```

Comment:
I'm not sure how to validate
```ts
[key: number]: {
    comments: Comment | Comment[];
};
```

`curl https://bugzilla.mozilla.org/rest/bug/comment/1` - API returns something like this:

```json
{
    "comments": {
        "1": {
            "time": "1998-04-07T08:37:03Z",
            "creation_time": "1998-04-07T08:37:03Z",
            "text": "Created by   (weitsang@cs.cornell.edu) on Monday, April 6, 1998 6:37:03 PM PDT\nAdditional Details :\nPREF_Cleanup is not called when Navigator exits\n(File->Exit), causing memory leaks.\nUpdated by Sarah Wilson (swilson@netscape.com) on Tuesday, April 7, 1998 6:48:16 PM PDT\nUpdated by Sarah Wilson (swilson@netscape.com) on Tuesday, April 7, 1998 6:49:27 PM PDT",
            "raw_text": "Created by   (weitsang@cs.cornell.edu) on Monday, April 6, 1998 6:37:03 PM PDT\nAdditional Details :\nPREF_Cleanup is not called when Navigator exits\n(File->Exit), causing memory leaks.\nUpdated by Sarah Wilson (swilson@netscape.com) on Tuesday, April 7, 1998 6:48:16 PM PDT\nUpdated by Sarah Wilson (swilson@netscape.com) on Tuesday, April 7, 1998 6:49:27 PM PDT",
            "attachment_id": null,
            "bug_id": 35,
            "is_private": false,
            "id": 1,
            "tags": [],
            "creator": "weitsang@cs.cornell.edu",
            "count": 0,
            "author": "weitsang@cs.cornell.edu"
        }
    },
    "bugs": {}
}
```

`curl https://bugzilla.redhat.com/rest/bug/1234567/comment`

```json
{
    "bugs": {
        "1234567": {
            "comments": [{
                "id": 8360743,
                "text": "The Emacs add-on packaging guidelines no longer stipulate that packages which also bundle support for Emacs should split out those Emacs files into separate sub-packages. This package should instead ship those files with the main package which should also Require emacs-filesystem. See https://fedoraproject.org/wiki/Packaging:Emacs for more detail.",
                "bug_id": 1234567,
                "tags": [],
                "is_private": false,
                "attachment_id": null,
                "count": 0,
                "creator_id": 163554,
                "time": "2015-06-22T17:53:39Z",
                "creation_time": "2015-06-22T17:53:39Z",
                "creator": "jonathan.underwood"
            }]
        }
    },
    "comments": {}
}
```

Reply:
Strange format. Untested but this looks about right:
```typescript
export function intString(val: any): number {
  if (typeof val != "string") {
    throw new Error(
      `Expected an integer as a string but received ${repr(val)}`,
    );
  }

  let value = parseInt(val, 10);
  if (value.toString() != val) {
    throw new Error(
      `Expected an integer as a string but received ${repr(val)}`,
    );
  }

  return value;
}

export function map<K, V>(
  keyValidator: Validator<K>,
  valueValidator: Validator<V>,
): Validator<Map<K, V>> {
  return (val: any): Map<K, V> => {
    if (!val || typeof val != "object") {
      throw new Error(`Expected an object but received ${repr(val)}`);
    }

    let result = new Map<K, V>();

    for (let [key, value] of Object.entries(val)) {
      result.set(keyValidator(key), valueValidator(value));
    }

    return result;
  };
}
```

Then you'd be able to use `map(intString, object(CommentsSpec))` for that first case as an example..

Reply:
Thank you @Mossop!!! Worked out of the box. I just updated `[key: number]: Comment` to `Map<number, Comment>`.